### PR TITLE
Raft

### PR DESCRIPTION
**Pull request by shayanh**
Raft basic protocol implementation.

This PR contains:
- Raft model checked spec.
- Raft Go tests.
- Locally shared archetype resourced implemented. This resource can be safely shared between concurrent running archetypes.
- Fix a bug in `mpcalctx.go`. Moved panic recovery before `ctx.preRun()` to catch `preRun()` panics. The previous version panic recovered by a monitor and caused infinite blocking on a subsequent `ctx.Stop()` call.
- TLA bug fix: According to TLC semantics, in `SubSeq` function if `from` is greater than `to`, just return an empty sequence.
- TLA bug fix: When making a TLA function, if domain of the function is not a tuple, then the output shouldn't be a tuple too.

### CODE REVIEW

File: distsys/resources/localshared.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		if resCh != nil {
+			<-resCh
+		}
+	}
+	if res.acquired > 0 {
+		res.sharedRes.release(res.acquired)
+		res.acquired = 0
+	}
+	ch <- struct{}{}
+	return ch
```
Comment (fhackett) [inline distsys/resources/localshared.go:80]: if you do all the work synchronously, you can just return `nil`, no need to allocate / manage a temporary channel

File: distsys/resources/localshared.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	acquired int64
+}
+
+func (res *LocalShared) Abort() chan struct{} {
+	ch := make(chan struct{}, 1)
+
+	if res.acquired == maxSemSize {
+		resCh := res.sharedRes.res.Abort()
+		if resCh != nil {
+			<-resCh
```
Comment (fhackett) [inline distsys/resources/localshared.go:72]: and, you can just return this directly I think, without even checking for `nil`

File: distsys/resources/localshared.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+		}
+	}
+	if res.acquired > 0 {
+		res.sharedRes.release(res.acquired)
+		res.acquired = 0
+	}
+	ch <- struct{}{}
+	return ch
```
Comment (fhackett) [inline distsys/resources/localshared.go:101]: same as for Abort

**Review by fhackett — COMMENTED** [review]
Aside from some minor details, this makes sense.

One concern going forward: I notice you're using an in-memory log here. We should have durable storage for Raft logs during eval, because from what I can tell etcd will not work without disk storage, and it would be an issue comparing the two meaningfully.
I can maybe look into that if you need to focus on 538B for a bit, now that this PR makes a starting / synchronisation point clear.

File: distsys/resources/localshared.go (hunk 1 above)
Comment (shayanh) [reply distsys/resources/localshared.go:80]: Initially, I implemented it async but for debugging I changed it to sync and then forgot to change it. Thanks for reminding me!

File: distsys/resources/localshared.go (hunk 2 above)
Comment (shayanh) [reply distsys/resources/localshared.go:72]: I want to release the lock when I know `sharedRes`'s abort is done.

**Comment by shayanh** [comment]
I agree that the persistence is required (according to Raft paper, in addition to `log`, we should also persist `currentTerm` and `votedFor`). This PR is so large and I suggest working on persistence in another PR. And yes, your help is much appreciated. I should work on 538B presentation and the writing.

**Comment by shayanh** [comment]
Also we can run etcd using tmpfs, so it uses memory as file system and there will no disk overhead. We can use this hack in case of need.

File: distsys/resources/localshared.go (hunk 2 above)
Comment (fhackett) [reply distsys/resources/localshared.go:70]: ah, I didn't see that.

**Comment by fhackett** [comment]
Re: tmpfs; I thought about it, but am not seriously considering it (yet). In case of need, yes.

I forgot one last comment: we should make sure the two TLA+ bugs are fixed in the interpreter as well, otherwise the fuzz tests will "find" the changes here eventually.

Otherwise, looks good.

**Pull request merged**